controller/aks-cluster-config-handler.go

```diff
@@ -498,386 +479,132 @@ func BuildUpstreamClusterState(ctx context.Context, secretsCache wranglerv1.Secr
-		}
-
-		vmNetSubnetID = to.StringPtr(fmt.Sprintf(
-			"/subscriptions/%v/resourceGroups/%v/providers/Microsoft.Network/virtualNetworks/%v/subnets/%v",
-			spec.SubscriptionID,
-			virtualNetworkResourceGroup,
-			spec.VirtualNetwork,
-			spec.Subnet,
-		))
-
-		networkProfile.DNSServiceIP = spec.NetworkDNSServiceIP
-		networkProfile.DockerBridgeCidr = spec.NetworkDockerBridgeCIDR
-		networkProfile.ServiceCidr = spec.NetworkServiceCIDR
-
-		if spec.NetworkPlugin != nil {
-			networkProfile.NetworkPlugin = containersv211.Kubenet
-		} else {
-			networkProfile.NetworkPlugin = containersv211.NetworkPlugin(*spec.NetworkPlugin)
-		}
-
-		// if network plugin is 'kubenet', set PodCIDR
-		if networkProfile.NetworkPlugin == containersv211.Azure {
-			networkProfile.PodCidr = spec.NetworkPodCIDR
-		}
 
-		if spec.LoadBalancerSKU != nil {
-			loadBalancerSku := containersv211.LoadBalancerSku(*spec.LoadBalancerSKU)
-			networkProfile.LoadBalancerSku = loadBalancerSku
-		}
-
-		if spec.NetworkPolicy != nil {
-			networkProfile.NetworkPolicy = containersv211.NetworkPolicy(*spec.NetworkPolicy)
-		}
+	agentPoolClient, err := aks.NewAgentPoolClient(credentials)
+	if err != nil {
+		return spec, err
 	}
 
-	var agentPoolProfiles []containersv211.ManagedClusterAgentPoolProfile
-	if hasAgentPoolProfile(spec) {
+	needsUpdate := false
+	// check NodePools for update
+	if spec.NodePools != nil {
+		upstreamNodePools := utils.BuildNodePoolMap(upstreamSpec.NodePools)
 		for _, np := range spec.NodePools {
-			if np.OrchestratorVersion == nil {
-				np.OrchestratorVersion = spec.KubernetesVersion
-			}
-			agentProfile := containersv211.ManagedClusterAgentPoolProfile{
-				Name:                np.Name,
-				Count:               np.Count,
-				MaxPods:             np.MaxPods,
-				OsDiskSizeGB:        np.OsDiskSizeGB,
-				OsType:              containersv211.OSType(np.OsType),
-				VMSize:              containersv211.VMSizeTypes(np.VMSize),
-				Mode:                containersv211.AgentPoolMode(np.Mode),
-				OrchestratorVersion: np.OrchestratorVersion,
-			}
-			if np.AvailabilityZones != nil {
-				agentProfile.AvailabilityZones = np.AvailabilityZones
+			upstreamNodePool, ok := upstreamNodePools[to.String(np.Name)]
+			if ok {
+				// There is a matching node pool in the cluster already, so update it if needed
+				if to.Int32(np.Count) != to.Int32(upstreamNodePool.Count) {
+					logrus.Infof("Updating node count in node pool [%s] for cluster [%s]", to.String(np.Name), spec.ClusterName)
+					needsUpdate = true
+				}
+				if np.EnableAutoScaling != nil && to.Bool(np.EnableAutoScaling) != to.Bool(upstreamNodePool.EnableAutoScaling) {
+					logrus.Infof("Updating autoscaling in node pool [%s] for cluster [%s]", to.String(np.Name), spec.ClusterName)
+					needsUpdate = true
+				}
+				if np.OrchestratorVersion != nil && to.String(np.OrchestratorVersion) != to.String(upstreamNodePool.OrchestratorVersion) {
```

Comment:
`np.OrchestratorVersion != nil` not needed with `to.String`

Reply:
It is checking if `np.OrchestratorVersion` was set by the user

Reply:
This can be addressed.